ES Index Out of Bounds Exception when Querying

**Elasticsearch version** (`bin/elasticsearch --version`): 7.9.2, 7.15.1

**Plugins installed**: None

**JVM version** (`java -version`): Built-in with RPM (for both versions)

**OS version** (`uname -a` if on a Unix-like system): CentOS 7, `Linux localhost 3.10.0-862.14.4.el7.x86_64 #1 SMP Wed Sep 26 15:12:11 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux`

**Description**:

# ES Index Out of Bounds Exception when Querying

We have a system that involves heavily nested data that includes possible (partially) repeated items on lower levels.  We have a 
query that counts (or sums) specific fields for the nested lower level data and compares the resulting value to a threshold to determine whether the document itself matches.  This query sometimes returns incorrect results or even blows up with an exception. We believe this to be a bug within Elasticsearch or Lucene.  

A simplified version of the query is illustrated below.  We use a match phrase query to choose the specific fields and use a constant score query with a value of 1 (for the counting case) for a score, and use a scoring mode of `sum` in the nested query to find the number of matches.  We then use a script with a `min_score` equal to threshold results so we only find counts exceeding the threshold.

This query usually works fine, but when the match phrase query is long we either:

1. Get an exception from ES with the message:
```
ElasticsearchException[Elasticsearch exception [type=index_out_of_bounds_exception, reason=index_out_of_bounds_exception: null]]
com.pfi.trex.enterprisemanager.search.es.control.ESSearchException: Search Failure  (0) ElasticsearchException[Elasticsearch exception [type=index_out_of_bounds_exception, reason=index_out_of_bounds_exception: null]]
```

2. Get inconsistent results where the results are either not found or only partial results are returned.

If we use a lower level query different from a match phrase query (e.g. a `match` query, or a `wildcard` query or a `fuzzy` match), then results are returned as expected.  Also (oddly) if we run the match phrase query with `profiling` set to `true`, it does return the expected results.

We have found this on a `7.9.2` version of Elasticsearch but have reproduced it on a system using `7.15.1` as well. We have seen this on a number of systems of various sizes. Nodes all run on Linux, typically CentOS 7 with the RPM distribution from Elastic.

In the following, we give (1) the stack trace that results from the index out of bounds exception, and then (2) provide a simplified example that shows the inconsistent results.

We only see the stack trace on a larger system with many items, but the simpified case at least shows that the query is not functioning properly.  The simplified query that we give below gives the same stack trace when run against a fully populated system.  The `size` or the `from` and `size` fields sometimes need adjustment in order to trigger the stack trace, depending upon the system.  Using a `size` of `0` also triggers the stack trace on the populated systems.

----------------------------------------------------------------

The Stack Trace (from a development system with log level debug):

```
[2021-10-14T13:15:24,068][INFO ][o.e.m.j.JvmGcMonitorService] [martyes] [gc][1892293] overhead, spent [261ms] collecting in the last [1s]
[2021-10-14T15:42:19,895][WARN ][r.suppressed             ] [martyes] path: /transitextmeta/_search, params: {typed_keys=true, max_concurrent_shard_requests=36, ignore_unavailable=false, expand_wildcards=open, allow_no_indices=true, ignore_throttled=true, index=transitextmeta, search_type=query_then_fetch, batched_reduce_size=512, ccs_minimize_roundtrips=true}
org.elasticsearch.action.search.SearchPhaseExecutionException: all shards failed
	at org.elasticsearch.action.search.AbstractSearchAsyncAction.onPhaseFailure(AbstractSearchAsyncAction.java:551) [elasticsearch-7.9.2.jar:7.9.2]
	at org.elasticsearch.action.search.AbstractSearchAsyncAction.executeNextPhase(AbstractSearchAsyncAction.java:309) [elasticsearch-7.9.2.jar:7.9.2]
	at org.elasticsearch.action.search.AbstractSearchAsyncAction.onPhaseDone(AbstractSearchAsyncAction.java:582) [elasticsearch-7.9.2.jar:7.9.2]
	at org.elasticsearch.action.search.AbstractSearchAsyncAction.onShardFailure(AbstractSearchAsyncAction.java:393) [elasticsearch-7.9.2.jar:7.9.2]
	at org.elasticsearch.action.search.AbstractSearchAsyncAction.access$100(AbstractSearchAsyncAction.java:68) [elasticsearch-7.9.2.jar:7.9.2]
	at org.elasticsearch.action.search.AbstractSearchAsyncAction$1.onFailure(AbstractSearchAsyncAction.java:245) [elasticsearch-7.9.2.jar:7.9.2]
	at org.elasticsearch.action.search.SearchExecutionStatsCollector.onFailure(SearchExecutionStatsCollector.java:73) [elasticsearch-7.9.2.jar:7.9.2]
	at org.elasticsearch.action.ActionListenerResponseHandler.handleException(ActionListenerResponseHandler.java:59) [elasticsearch-7.9.2.jar:7.9.2]
	at org.elasticsearch.action.search.SearchTransportService$ConnectionCountingHandler.handleException(SearchTransportService.java:403) [elasticsearch-7.9.2.jar:7.9.2]
	at org.elasticsearch.transport.TransportService$6.handleException(TransportService.java:638) [elasticsearch-7.9.2.jar:7.9.2]
	at org.elasticsearch.transport.TransportService$ContextRestoreResponseHandler.handleException(TransportService.java:1172) [elasticsearch-7.9.2.jar:7.9.2]
	at org.elasticsearch.transport.InboundHandler.lambda$handleException$2(InboundHandler.java:235) [elasticsearch-7.9.2.jar:7.9.2]
	at org.elasticsearch.common.util.concurrent.EsExecutors$DirectExecutorService.execute(EsExecutors.java:226) [elasticsearch-7.9.2.jar:7.9.2]
	at org.elasticsearch.transport.InboundHandler.handleException(InboundHandler.java:233) [elasticsearch-7.9.2.jar:7.9.2]
	at org.elasticsearch.transport.InboundHandler.handlerResponseError(InboundHandler.java:225) [elasticsearch-7.9.2.jar:7.9.2]
	at org.elasticsearch.transport.InboundHandler.messageReceived(InboundHandler.java:115) [elasticsearch-7.9.2.jar:7.9.2]
	at org.elasticsearch.transport.InboundHandler.inboundMessage(InboundHandler.java:78) [elasticsearch-7.9.2.jar:7.9.2]
	at org.elasticsearch.transport.TcpTransport.inboundMessage(TcpTransport.java:692) [elasticsearch-7.9.2.jar:7.9.2]
	at org.elasticsearch.transport.InboundPipeline.forwardFragments(InboundPipeline.java:142) [elasticsearch-7.9.2.jar:7.9.2]
	at org.elasticsearch.transport.InboundPipeline.doHandleBytes(InboundPipeline.java:117) [elasticsearch-7.9.2.jar:7.9.2]
	at org.elasticsearch.transport.InboundPipeline.handleBytes(InboundPipeline.java:82) [elasticsearch-7.9.2.jar:7.9.2]
	at org.elasticsearch.transport.netty4.Netty4MessageChannelHandler.channelRead(Netty4MessageChannelHandler.java:76) [transport-netty4-client-7.9.2.jar:7.9.2]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379) [netty-transport-4.1.49.Final.jar:4.1.49.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365) [netty-transport-4.1.49.Final.jar:4.1.49.Final]
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357) [netty-transport-4.1.49.Final.jar:4.1.49.Final]
	at io.netty.handler.logging.LoggingHandler.channelRead(LoggingHandler.java:271) [netty-handler-4.1.49.Final.jar:4.1.49.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379) [netty-transport-4.1.49.Final.jar:4.1.49.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365) [netty-transport-4.1.49.Final.jar:4.1.49.Final]
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357) [netty-transport-4.1.49.Final.jar:4.1.49.Final]
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410) [netty-transport-4.1.49.Final.jar:4.1.49.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379) [netty-transport-4.1.49.Final.jar:4.1.49.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365) [netty-transport-4.1.49.Final.jar:4.1.49.Final]
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919) [netty-transport-4.1.49.Final.jar:4.1.49.Final]
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:163) [netty-transport-4.1.49.Final.jar:4.1.49.Final]
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:714) [netty-transport-4.1.49.Final.jar:4.1.49.Final]
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysPlain(NioEventLoop.java:615) [netty-transport-4.1.49.Final.jar:4.1.49.Final]
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:578) [netty-transport-4.1.49.Final.jar:4.1.49.Final]
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:493) [netty-transport-4.1.49.Final.jar:4.1.49.Final]
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989) [netty-common-4.1.49.Final.jar:4.1.49.Final]
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74) [netty-common-4.1.49.Final.jar:4.1.49.Final]
	at java.lang.Thread.run(Thread.java:832) [?:?]
Caused by: org.elasticsearch.common.io.stream.NotSerializableExceptionWrapper: index_out_of_bounds_exception: null
	at java.nio.Buffer.checkIndex(Buffer.java:744) ~[?:?]
	at java.nio.DirectByteBuffer.getLong(DirectByteBuffer.java:920) ~[?:?]
	at org.apache.lucene.store.ByteBufferGuard.getLong(ByteBufferGuard.java:134) ~[lucene-core-8.6.2.jar:8.6.2 016993b65e393b58246d54e8ddda9f56a453eb0e - ivera - 2020-08-26 10:53:36]
	at org.apache.lucene.store.ByteBufferIndexInput$SingleBufferImpl.readLong(ByteBufferIndexInput.java:479) ~[lucene-core-8.6.2.jar:8.6.2 016993b65e393b58246d54e8ddda9f56a453eb0e - ivera - 2020-08-26 10:53:36]
	at org.apache.lucene.util.packed.DirectReader$DirectPackedReader64.get(DirectReader.java:343) ~[lucene-core-8.6.2.jar:8.6.2 016993b65e393b58246d54e8ddda9f56a453eb0e - ivera - 2020-08-26 10:53:36]
	at org.apache.lucene.codecs.lucene80.Lucene80DocValuesProducer$8.longValue(Lucene80DocValuesProducer.java:545) ~[lucene-core-8.6.2.jar:8.6.2 016993b65e393b58246d54e8ddda9f56a453eb0e - ivera - 2020-08-26 10:53:36]
	at org.apache.lucene.document.SortedNumericDocValuesRangeQuery$1$1.matches(SortedNumericDocValuesRangeQuery.java:124) ~[lucene-core-8.6.2.jar:8.6.2 016993b65e393b58246d54e8ddda9f56a453eb0e - ivera - 2020-08-26 10:53:36]
	at org.apache.lucene.search.ConjunctionDISI$ConjunctionTwoPhaseIterator.matches(ConjunctionDISI.java:345) ~[lucene-core-8.6.2.jar:8.6.2 016993b65e393b58246d54e8ddda9f56a453eb0e - ivera - 2020-08-26 10:53:36]
	at org.apache.lucene.search.join.ToParentBlockJoinQuery$ParentTwoPhase.matches(ToParentBlockJoinQuery.java:262) ~[lucene-join-8.6.2.jar:8.6.2 016993b65e393b58246d54e8ddda9f56a453eb0e - ivera - 2020-08-26 10:54:38]
	at org.apache.lucene.search.ConjunctionDISI$ConjunctionTwoPhaseIterator.matches(ConjunctionDISI.java:345) ~[lucene-core-8.6.2.jar:8.6.2 016993b65e393b58246d54e8ddda9f56a453eb0e - ivera - 2020-08-26 10:53:36]
	at org.apache.lucene.search.Weight$DefaultBulkScorer.scoreRange(Weight.java:250) ~[lucene-core-8.6.2.jar:8.6.2 016993b65e393b58246d54e8ddda9f56a453eb0e - ivera - 2020-08-26 10:53:36]
	at org.apache.lucene.search.Weight$DefaultBulkScorer.score(Weight.java:229) ~[lucene-core-8.6.2.jar:8.6.2 016993b65e393b58246d54e8ddda9f56a453eb0e - ivera - 2020-08-26 10:53:36]
	at org.elasticsearch.search.internal.CancellableBulkScorer.score(CancellableBulkScorer.java:56) ~[elasticsearch-7.9.2.jar:7.9.2]
	at org.apache.lucene.search.BulkScorer.score(BulkScorer.java:39) ~[lucene-core-8.6.2.jar:8.6.2 016993b65e393b58246d54e8ddda9f56a453eb0e - ivera - 2020-08-26 10:53:36]
	at org.elasticsearch.search.internal.ContextIndexSearcher.searchLeaf(ContextIndexSearcher.java:226) ~[elasticsearch-7.9.2.jar:7.9.2]
	at org.elasticsearch.search.internal.ContextIndexSearcher.search(ContextIndexSearcher.java:199) ~[elasticsearch-7.9.2.jar:7.9.2]
	at org.apache.lucene.search.IndexSearcher.search(IndexSearcher.java:445) ~[lucene-core-8.6.2.jar:8.6.2 016993b65e393b58246d54e8ddda9f56a453eb0e - ivera - 2020-08-26 10:53:36]
	at org.elasticsearch.search.query.QueryPhase.searchWithCollector(QueryPhase.java:343) ~[elasticsearch-7.9.2.jar:7.9.2]
	at org.elasticsearch.search.query.QueryPhase.executeInternal(QueryPhase.java:298) ~[elasticsearch-7.9.2.jar:7.9.2]
	at org.elasticsearch.search.query.QueryPhase.execute(QueryPhase.java:150) ~[elasticsearch-7.9.2.jar:7.9.2]
	at org.elasticsearch.search.SearchService.loadOrExecuteQueryPhase(SearchService.java:362) ~[elasticsearch-7.9.2.jar:7.9.2]
	at org.elasticsearch.search.SearchService.executeQueryPhase(SearchService.java:435) ~[elasticsearch-7.9.2.jar:7.9.2]
	at org.elasticsearch.search.SearchService.access$200(SearchService.java:136) ~[elasticsearch-7.9.2.jar:7.9.2]
	at org.elasticsearch.search.SearchService$2.lambda$onResponse$0(SearchService.java:396) ~[elasticsearch-7.9.2.jar:7.9.2]
	at org.elasticsearch.search.SearchService.lambda$runAsync$0(SearchService.java:412) ~[elasticsearch-7.9.2.jar:7.9.2]
	at org.elasticsearch.common.util.concurrent.TimedRunnable.doRun(TimedRunnable.java:44) ~[elasticsearch-7.9.2.jar:7.9.2]
	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:737) ~[elasticsearch-7.9.2.jar:7.9.2]
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37) ~[elasticsearch-7.9.2.jar:7.9.2]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630) ~[?:?]
	at java.lang.Thread.run(Thread.java:832) ~[?:?]
```

Example error return when running the sample query against a fully populated system:
```
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  4565  100   932  100  3633    242    945  0:00:03  0:00:03 --:--:--  1187{
  "error" : {
    "root_cause" : [
      {
        "type" : "index_out_of_bounds_exception",
        "reason" : "index_out_of_bounds_exception: null"
      },
      {
        "type" : "index_out_of_bounds_exception",
        "reason" : null
      }
    ],
    "type" : "search_phase_execution_exception",
    "reason" : "all shards failed",
    "phase" : "query",
    "grouped" : true,
    "failed_shards" : [
      {
        "shard" : 0,
        "index" : "[redacted]",
        "node" : "VeddA7VTTz-oqvvyVECOsQ",
        "reason" : {
          "type" : "index_out_of_bounds_exception",
          "reason" : "index_out_of_bounds_exception: null"
        }
      },
      {
        "shard" : 1,
        "index" : "[redacted]",
        "node" : "dYjvEQmLSi6CeL4VVElGKw",
        "reason" : {
          "type" : "index_out_of_bounds_exception",
          "reason" : null
        }
      }
    ]
  },
  "status" : 500
}
```


----------------------------------------------------------------

To Reproduce the Inconsistent Behavior:

This illustrates that the normal `match_phrase` query fails for this simple case, but that with profiling enabled, or with a different low level query, it succeeds.  Both the index and queries have been greatly simplified compared to what is run on the deployed systems.  In the following, please substitute your ES node IP address for what is used in the following.  This was run against a dev system with more than one node, so there were two shards created, the primary and a replica.

1. Create an index with the following mapping:

```
$ curl -X PUT "localhost:9200/outofboundsv1?pretty" -H 'Content-Type: application/json' -d'
{
  "aliases": {
    "outofbounds": {}
  },
  "mappings": {
    "dynamic": "strict",
    "properties": {
      "toolTracks": {
        "type": "nested",
        "properties": {
          "idToken": {
            "type": "text"
          },
          "data": {
            "type": "text"
          },
          "start" : {
            "type" : "long"
          },
          "confidence": {
            "type": "double"
          }
        }
      },
      "md5": {
        "type": "keyword"
      },
      "durationMSecs": {
        "type": "long"
      },
      "calculatedDurationMSecs": {
        "type": "long"
      }
    }
  },
  "settings": {
    "index": {
      "requests.cache.enable": true,
      "max_result_window": 1000000,
      "max_inner_result_window": 1000000,
      "refresh_interval": "30s"
    }
  }
}
'
```
Results:
```
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   916  100    89  100   827     43    408  0:00:02  0:00:02 --:--:--   452{
  "acknowledged" : true,
  "shards_acknowledged" : true,
  "index" : "outofboundsv1"
}
```

2. Populate it with a single document with several records of a nested type:

```
$ curl -X POST "localhost:9200/outofboundsv1/_doc/00001?pretty" -H 'Content-Type: application/json' -d'
{
  "md5" : "00001",
  "durationMSecs" : "2000",
  "toolTracks": [{
	"idToken" : "videoLabelsTemporalID",
         "start" : 10,
         "data" : "cash machine, cash dispenser, automated teller machine, automatic teller machine, automated teller, automatic teller, ATM",
         "confidence" : 0.3
  },
  {
	"idToken" : "videoLabelsTemporalID",
         "start" : 100,
         "data" : "cash machine, cash dispenser, automated teller machine, automatic teller machine, automated teller, automatic teller, ATM",
         "confidence" : 0.92
  },
  {
	"idToken" : "videoLabelsTemporalID",
         "start" : 200,
         "data" : "cash machine, cash dispenser, automated teller machine, automatic teller machine, automated teller, automatic teller, ATM",
         "confidence" : 0.7
  },
  {
	"idToken" : "videoLabelsTemporalID",
         "start" : 300,
         "data" : "cash machine, cash dispenser, automated teller machine, automatic teller machine, automated teller, automatic teller, ATM",
         "confidence" : 0.85
  },
  {
	"idToken" : "videoLabelsTemporalID",
         "start" : 400,
         "data" : "cash machine, cash dispenser, automated teller machine, automatic teller machine, automated teller, automatic teller, ATM",
         "confidence" : 0.2
  },
  {
	"idToken" : "videoLabelsTemporalID",
         "start" : 500,
         "data" : "cash machine, cash dispenser, automated teller machine, automatic teller machine, automated teller, automatic teller, ATM",
         "confidence" : 0.3
  },
  {
	"idToken" : "videoLabelsTemporalID",
         "start" : 600,
         "data" : "cash machine, cash dispenser, automated teller machine, automatic teller machine, automated teller, automatic teller, ATM",
         "confidence" : 0.75
  },
  {
	"idToken" : "videoLabelsTemporalID",
         "start" : 700,
         "data" : "cash machine, cash dispenser, automated teller machine, automatic teller machine, automated teller, automatic teller, ATM",
         "confidence" : 0.82
  },
  {
	"idToken" : "videoLabelsTemporalID",
         "start" : 800,
         "data" : "cash machine, cash dispenser, automated teller machine, automatic teller machine, automated teller, automatic teller, ATM",
         "confidence" : 0.1
  },
  {
	"idToken" : "videoLabelsTemporalID",
         "start" : 900,
         "data" : "cash machine, cash dispenser, automated teller machine, automatic teller machine, automated teller, automatic teller, ATM",
         "confidence" : 0.6
  },
  {
	"idToken" : "videoLabelsTemporalID",
         "start" : 1000,
         "data" : "cash machine, cash dispenser, automated teller machine, automatic teller machine, automated teller, automatic teller, ATM",
         "confidence" : 0.3
  },
  {
	"idToken" : "videoLabelsTemporalID",
         "start" : 1100,
         "data" : "cash machine, cash dispenser, automated teller machine, automatic teller machine, automated teller, automatic teller, ATM",
         "confidence" : 0.7
  }
 ]
}
'
```

Results:

```
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  3041  100   231  100  2810   1358  16529 --:--:-- --:--:-- --:--:-- 17888{
  "_index" : "outofboundsv1",
  "_type" : "_doc",
  "_id" : "00001",
  "_version" : 1,
  "result" : "created",
  "_shards" : {
    "total" : 2,
    "successful" : 2,
    "failed" : 0
  },
  "_seq_no" : 0,
  "_primary_term" : 1
}
```

3. To validate the query, use a `match_phrase` Query with profiling ENABLED.  This returns the valid result; it does find the document.  Aside from profiling, this query is identical to the failing version.

```
$ curl -X POST "localhost:9200/outofboundsv1/_search/?pretty" -H 'Content-Type: application/json' -d'
{
  "from": 0,
  "size": 200,
  "timeout": "600000ms",
  "profile" : true,
  "query": {
    "bool": {
      "filter": [
                    {
                      "script_score": {
                        "query": {
                          "nested": {
                            "query": {
                              "constant_score": {
                                "filter": {
                                  "bool": {
                                    "filter": [
                                      {
                                        "match_phrase": {
                                          "toolTracks.data": {
                                            "query": "cash machine, cash dispenser, automated teller machine, automatic teller machine, automated teller, automatic teller, atm",
                                            "slop": 0,
                                            "zero_terms_query": "NONE",
                                            "boost": 1.0
                                          }
                                        }
                                      },
                                      {
                                        "match": {
                                          "toolTracks.idToken": {
                                            "query": "videoLabelsTemporalID",
                                            "operator": "OR",
                                            "prefix_length": 0,
                                            "max_expansions": 50,
                                            "fuzzy_transpositions": true,
                                            "lenient": false,
                                            "zero_terms_query": "NONE",
                                            "auto_generate_synonyms_phrase_query": true,
                                            "boost": 1.0
                                          }
                                        }
                                      },
                                      {
                                        "range": {
                                           "toolTracks.confidence": {
                                            "from": 0.8,
                                            "to": null,
                                            "include_lower": true,
                                            "include_upper": true,
                                            "boost": 1.0
                                          }
                                        }
                                      }
                                    ],
                                    "adjust_pure_negative": true,
                                    "boost": 1.0
                                  }
                                },
                                "boost": 1.0
                              }
                            },
                            "path": "toolTracks",
                            "ignore_unmapped": false,
                            "score_mode": "sum",
                            "boost": 1.0
                          }
                        },
                        "script": {
                          "source": "_score",
                          "lang": "painless"
                        },
                        "min_score": 2.0,
                        "boost": 1.0
                      }
                    }
                  ],
      "adjust_pure_negative": true,
      "boost": 1.0
    }
  },
  "_source": false,
  "track_total_hits": 2147483647
}'
```

Results (hits.total.value = 1):

```
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0{
  "took" : 3,
  "timed_out" : false,
  "_shards" : {
    "total" : 1,
    "successful" : 1,
    "skipped" : 0,
    "failed" : 0
  },
  "hits" : {
    "total" : {
      "value" : 1,
      "relation" : "eq"
    },
    "max_score" : 0.0,
    "hits" : [
      {
        "_index" : "outofboundsv1",
        "_type" : "_doc",
        "_id" : "00001",
        "_score" : 0.0
      }
    ]
  },
  "profile" : { ... } //suppressing display of profile

```

4. The failing case: use a `match_phrase` Query but for normal use with profiling DISABLED.  This finds no hits as shown below.

```
$ curl -X POST "localhost:9200/outofboundsv1/_search/?pretty" -H 'Content-Type: application/json' -d'
{
  "from": 0,
  "size": 200,
  "timeout": "600000ms",
  "profile" : false,
  "query": {
    "bool": {
      "filter": [
                    {
                      "script_score": {
                        "query": {
                          "nested": {
                            "query": {
                              "constant_score": {
                                "filter": {
                                  "bool": {
                                    "filter": [
                                      {
                                        "match_phrase": {
                                          "toolTracks.data": {
                                            "query": "cash machine, cash dispenser, automated teller machine, automatic teller machine, automated teller, automatic teller, atm",
                                            "slop": 0,
                                            "zero_terms_query": "NONE",
                                            "boost": 1.0
                                          }
                                        }
                                      },
                                      {
                                        "match": {
                                          "toolTracks.idToken": {
                                            "query": "videoLabelsTemporalID",
                                            "operator": "OR",
                                            "prefix_length": 0,
                                            "max_expansions": 50,
                                            "fuzzy_transpositions": true,
                                            "lenient": false,
                                            "zero_terms_query": "NONE",
                                            "auto_generate_synonyms_phrase_query": true,
                                            "boost": 1.0
                                          }
                                        }
                                      },
                                      {
                                        "range": {
                                           "toolTracks.confidence": {
                                            "from": 0.8,
                                            "to": null,
                                            "include_lower": true,
                                            "include_upper": true,
                                            "boost": 1.0
                                          }
                                        }
                                      }
                                    ],
                                    "adjust_pure_negative": true,
                                    "boost": 1.0
                                  }
                                },
                                "boost": 1.0
                              }
                            },
                            "path": "toolTracks",
                            "ignore_unmapped": false,
                            "score_mode": "sum",
                            "boost": 1.0
                          }
                        },
                        "script": {
                          "source": "_score",
                          "lang": "painless"
                        },
                        "min_score": 2.0,
                        "boost": 1.0
                      }
                    }
                  ],
      "adjust_pure_negative": true,
      "boost": 1.0
    }
  },
  "_source": false,
  "track_total_hits": 2147483647
}'
```

Results (hits.total.value == 0)

```
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  3881  100   261  100  3620  43500   589k --:--:-- --:--:-- --:--:--  758k{
  "took" : 3,
  "timed_out" : false,
  "_shards" : {
    "total" : 1,
    "successful" : 1,
    "skipped" : 0,
    "failed" : 0
  },
  "hits" : {
    "total" : {
      "value" : 0,
      "relation" : "eq"
    },
    "max_score" : null,
    "hits" : [ ]
  }
}
```

5. Change the `match_phrase` query to a `match` query and see that it works:

```
$ curl -X POST "localhost:9200/outofboundsv1/_search/?pretty" -H 'Content-Type: application/json' -d'
{
  "from": 0,
  "size": 200,
  "timeout": "600000ms",
  "profile" : false,
  "query": {
    "bool": {
      "filter": [
                    {
                      "script_score": {
                        "query": {
                          "nested": {
                            "query": {
                              "constant_score": {
                                "filter": {
                                  "bool": {
                                    "filter": [
                                      {
                                        "match": {
                                          "toolTracks.data": {
                                            "query": "cash machine, cash dispenser, automated teller machine, automatic teller machine, automated teller, automatic teller, atm",
                                            "zero_terms_query": "NONE",
                                            "boost": 1.0
                                          }
                                        }
                                      },
                                      {
                                        "match": {
                                          "toolTracks.idToken": {
                                            "query": "videoLabelsTemporalID",
                                            "operator": "OR",
                                            "prefix_length": 0,
                                            "max_expansions": 50,
                                            "fuzzy_transpositions": true,
                                            "lenient": false,
                                            "zero_terms_query": "NONE",
                                            "auto_generate_synonyms_phrase_query": true,
                                            "boost": 1.0
                                          }
                                        }
                                      },
                                      {
                                        "range": {
                                           "toolTracks.confidence": {
                                            "from": 0.8,
                                            "to": null,
                                            "include_lower": true,
                                            "include_upper": true,
                                            "boost": 1.0
                                          }
                                        }
                                      }
                                    ],
                                    "adjust_pure_negative": true,
                                    "boost": 1.0
                                  }
                                },
                                "boost": 1.0
                              }
                            },
                            "path": "toolTracks",
                            "ignore_unmapped": false,
                            "score_mode": "sum",
                            "boost": 1.0
                          }
                        },
                        "script": {
                          "source": "_score",
                          "lang": "painless"
                        },
                        "min_score": 2.0,
                        "boost": 1.0
                      }
                    }
                  ],
      "adjust_pure_negative": true,
      "boost": 1.0
    }
  },
  "_source": false,
  "track_total_hits": 2147483647
}'
```
Results:

```
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  3948  100   390  100  3558  65000   579k --:--:-- --:--:-- --:--:--  642k{
  "took" : 3,
  "timed_out" : false,
  "_shards" : {
    "total" : 1,
    "successful" : 1,
    "skipped" : 0,
    "failed" : 0
  },
  "hits" : {
    "total" : {
      "value" : 1,
      "relation" : "eq"
    },
    "max_score" : 0.0,
    "hits" : [
      {
        "_index" : "outofboundsv1",
        "_type" : "_doc",
        "_id" : "00001",
        "_score" : 0.0
      }
    ]
  }
}
```


-------------------------------------------------------------
Other comments:
1. The mapping and query above are greatly simplified from the production version, but show the same behavior
2. If the `toolTracks.data` query text is shortened significantly in the match phrase above we do get hits.  But for other examples we haven't found a fixed size cutoff.  Also, in most cases with shorter query texts, the query works fine.
2. In some versions of our production queries, we use an inner hit stategy to return the lower level results.  So we would rather not completely restructure the query.  (This is not used for the version that gives the index out of bounds exception)
3. We could potentially work around the problem by using a `match` query instead of the `match_phrase` query, but that is obviously not as precise.

--- a/server/src/internalClusterTest/java/org/elasticsearch/search/nested/NestedWithMinScoreIT.java
+++ b/server/src/internalClusterTest/java/org/elasticsearch/search/nested/NestedWithMinScoreIT.java
@@ -1,0 +1,121 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.search.nested;
+
+import org.apache.lucene.search.join.ScoreMode;
+import org.elasticsearch.action.search.SearchRequest;
+import org.elasticsearch.action.search.SearchResponse;
+import org.elasticsearch.action.support.WriteRequest;
+import org.elasticsearch.index.query.BoolQueryBuilder;
+import org.elasticsearch.index.query.ConstantScoreQueryBuilder;
+import org.elasticsearch.index.query.MatchPhraseQueryBuilder;
+import org.elasticsearch.index.query.NestedQueryBuilder;
+import org.elasticsearch.index.query.RangeQueryBuilder;
+import org.elasticsearch.index.query.functionscore.ScriptScoreQueryBuilder;
+import org.elasticsearch.plugins.Plugin;
+import org.elasticsearch.script.MockScriptPlugin;
+import org.elasticsearch.script.Script;
+import org.elasticsearch.script.ScriptType;
+import org.elasticsearch.search.builder.SearchSourceBuilder;
+import org.elasticsearch.test.ESIntegTestCase;
+import org.elasticsearch.test.hamcrest.ElasticsearchAssertions;
+import org.elasticsearch.xcontent.XContentBuilder;
+import org.elasticsearch.xcontent.XContentFactory;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+
+public class NestedWithMinScoreIT extends ESIntegTestCase {
+
+    public static class ScriptTestPlugin extends MockScriptPlugin {
+        @Override
+        protected Map<String, Function<Map<String, Object>, Object>> pluginScripts() {
+            return Collections.singletonMap("score_script", params -> {
+                final Object scoreAccessor = params.get("_score");
+                if (scoreAccessor instanceof Number) {
+                    return ((Number) scoreAccessor).doubleValue();
+                } else {
+                    return null;
+                }
+            });
+        }
+    }
+
+    @Override
+    protected Collection<Class<? extends Plugin>> getMockPlugins() {
+        final List<Class<? extends Plugin>> plugins = new ArrayList<>(super.getMockPlugins());
+        plugins.add(ScriptTestPlugin.class);
+        return plugins;
+    }
+
+    public void testNestedWithMinScore() throws Exception {
+        XContentBuilder mapping = XContentFactory.jsonBuilder();
+        mapping.startObject();
+        mapping.startObject("properties");
+        {
+            mapping.startObject("toolTracks");
+            {
+                mapping.field("type", "nested");
+                mapping.startObject("properties");
+                {
+                    mapping.startObject("data");
+                    mapping.field("type", "text");
+                    mapping.endObject();
+
+                    mapping.startObject("confidence");
+                    mapping.field("type", "double");
+                    mapping.endObject();
+                }
+                mapping.endObject();
+            }
+            mapping.endObject();
+        }
+        mapping.endObject();
+        mapping.endObject();
+
+        client().admin().indices().prepareCreate("test").addMapping("_doc", mapping).get();
+
+        XContentBuilder doc = XContentFactory.jsonBuilder();
+        doc.startObject();
+        doc.startArray("toolTracks");
+        double[] confidence = new double[] { 0.3, 0.92, 0.7, 0.85, 0.2, 0.3, 0.75, 0.82, 0.1, 0.6, 0.3, 0.7 };
+        for (double v : confidence) {
+            doc.startObject();
+            doc.field("confidence", v);
+            doc.field("data", "cash dispenser, automated teller machine, automatic teller machine");
+            doc.endObject();
+        }
+        doc.endArray();
+        doc.endObject();
+
+        client().prepareIndex("test", "_doc").setId("d1").setRefreshPolicy(WriteRequest.RefreshPolicy.IMMEDIATE).setSource(doc).get();
+        final BoolQueryBuilder childQuery = new BoolQueryBuilder().filter(
+            new MatchPhraseQueryBuilder("toolTracks.data", "cash dispenser, automated teller machine, automatic teller machine")
+        ).filter(new RangeQueryBuilder("toolTracks.confidence").from(0.8));
+
+        final ScriptScoreQueryBuilder scriptScoreQuery = new ScriptScoreQueryBuilder(
+            new NestedQueryBuilder("toolTracks", new ConstantScoreQueryBuilder(childQuery), ScoreMode.Total),
+            new Script(ScriptType.INLINE, MockScriptPlugin.NAME, "score_script", Collections.emptyMap())
+        );
+        scriptScoreQuery.setMinScore(1.0f);
+        SearchSourceBuilder source = new SearchSourceBuilder();
+        source.query(scriptScoreQuery);
+        source.profile(randomBoolean());
+        if (randomBoolean()) {
+            source.trackTotalHitsUpTo(randomBoolean() ? Integer.MAX_VALUE : randomIntBetween(1, 1000));
+        }
+        SearchRequest searchRequest = new SearchRequest("test").source(source);
+        final SearchResponse searchResponse = client().search(searchRequest).actionGet();
+        ElasticsearchAssertions.assertSearchHits(searchResponse, "d1");
+    }
+}

--- a/server/src/main/java/org/elasticsearch/common/lucene/search/function/MinScoreScorer.java
+++ b/server/src/main/java/org/elasticsearch/common/lucene/search/function/MinScoreScorer.java
@@ -61,15 +61,25 @@ final class MinScoreScorer extends Scorer {
 
     @Override
     public TwoPhaseIterator twoPhaseIterator() {
-        final TwoPhaseIterator inTwoPhase = this.in.twoPhaseIterator();
-        final DocIdSetIterator approximation = inTwoPhase == null ? in.iterator() : inTwoPhase.approximation();
+        TwoPhaseIterator inTwoPhase = in.twoPhaseIterator();
+        DocIdSetIterator approximation;
+        if (inTwoPhase == null) {
+            approximation = in.iterator();
+            if (TwoPhaseIterator.unwrap(approximation) != null) {
+                inTwoPhase = TwoPhaseIterator.unwrap(approximation);
+                approximation = inTwoPhase.approximation();
+            }
+        } else {
+            approximation = inTwoPhase.approximation();
+        }
+        final TwoPhaseIterator finalTwoPhase = inTwoPhase;
         return new TwoPhaseIterator(approximation) {
 
             @Override
             public boolean matches() throws IOException {
                 // we need to check the two-phase iterator first
                 // otherwise calling score() is illegal
-                if (inTwoPhase != null && inTwoPhase.matches() == false) {
+                if (finalTwoPhase != null && finalTwoPhase.matches() == false) {
                     return false;
                 }
                 curScore = in.score();
@@ -79,7 +89,7 @@ final class MinScoreScorer extends Scorer {
             @Override
             public float matchCost() {
                 return 1000f // random constant for the score computation
-                    + (inTwoPhase == null ? 0 : inTwoPhase.matchCost());
+                    + (finalTwoPhase == null ? 0 : finalTwoPhase.matchCost());
             }
         };
     }

--- a/server/src/main/java/org/elasticsearch/common/lucene/search/function/ScriptScoreQuery.java
+++ b/server/src/main/java/org/elasticsearch/common/lucene/search/function/ScriptScoreQuery.java
@@ -23,6 +23,7 @@ import org.apache.lucene.search.QueryVisitor;
 import org.apache.lucene.search.Scorable;
 import org.apache.lucene.search.ScoreMode;
 import org.apache.lucene.search.Scorer;
+import org.apache.lucene.search.TwoPhaseIterator;
 import org.apache.lucene.search.Weight;
 import org.apache.lucene.util.Bits;
 import org.elasticsearch.Version;
@@ -263,6 +264,11 @@ public class ScriptScoreQuery extends Query {
         @Override
         public DocIdSetIterator iterator() {
             return subQueryScorer.iterator();
+        }
+
+        @Override
+        public TwoPhaseIterator twoPhaseIterator() {
+            return subQueryScorer.twoPhaseIterator();
         }
 
         @Override

--- a/server/src/test/java/org/elasticsearch/common/lucene/search/function/MinScoreScorerTests.java
+++ b/server/src/test/java/org/elasticsearch/common/lucene/search/function/MinScoreScorerTests.java
@@ -8,23 +8,36 @@
 
 package org.elasticsearch.common.lucene.search.function;
 
+import com.carrotsearch.randomizedtesting.generators.RandomPicks;
+
 import org.apache.lucene.index.LeafReaderContext;
 import org.apache.lucene.index.Term;
+import org.apache.lucene.search.AssertingScorer;
+import org.apache.lucene.search.ConjunctionDISI;
 import org.apache.lucene.search.DocIdSetIterator;
 import org.apache.lucene.search.Explanation;
 import org.apache.lucene.search.MatchAllDocsQuery;
+import org.apache.lucene.search.ScoreMode;
 import org.apache.lucene.search.Scorer;
 import org.apache.lucene.search.TwoPhaseIterator;
 import org.apache.lucene.search.Weight;
-import org.apache.lucene.util.LuceneTestCase;
 import org.apache.lucene.util.TestUtil;
+import org.elasticsearch.test.ESTestCase;
 
 import java.io.IOException;
+import java.util.ArrayList;
 import java.util.Arrays;
+import java.util.HashMap;
 import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
 import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
 
-public class MinScoreScorerTests extends LuceneTestCase {
+import static org.hamcrest.Matchers.equalTo;
+
+public class MinScoreScorerTests extends ESTestCase {
 
     private static DocIdSetIterator iterator(final int... docs) {
         return new DocIdSetIterator() {
@@ -33,11 +46,8 @@ public class MinScoreScorerTests extends LuceneTestCase {
 
             @Override
             public int nextDoc() throws IOException {
-                if (i + 1 == docs.length) {
-                    return NO_MORE_DOCS;
-                } else {
-                    return docs[++i];
-                }
+                ++i;
+                return docID();
             }
 
             @Override
@@ -81,11 +91,41 @@ public class MinScoreScorerTests extends LuceneTestCase {
         };
     }
 
+    private static Scorer hideTwoPhaseIterator(Scorer in) {
+        return new Scorer(in.getWeight()) {
+            @Override
+            public DocIdSetIterator iterator() {
+                return TwoPhaseIterator.asDocIdSetIterator(in.twoPhaseIterator());
+            }
+
+            @Override
+            public TwoPhaseIterator twoPhaseIterator() {
+                return null;
+            }
+
+            @Override
+            public float getMaxScore(int upTo) throws IOException {
+                return in.getMaxScore(upTo);
+            }
+
+            @Override
+            public float score() throws IOException {
+                return in.score();
+            }
+
+            @Override
+            public int docID() {
+                return in.docID();
+            }
+        };
+    }
+
     private static Scorer scorer(int maxDoc, final int[] docs, final float[] scores, final boolean twoPhase) {
         final DocIdSetIterator iterator = twoPhase ? DocIdSetIterator.all(maxDoc) : iterator(docs);
-        return new Scorer(fakeWeight()) {
+        final Scorer scorer = new Scorer(fakeWeight()) {
 
             int lastScoredDoc = -1;
+            final float matchCost = (random().nextBoolean() ? 1000 : 0) + random().nextInt(2000);
 
             public DocIdSetIterator iterator() {
                 if (twoPhase) {
@@ -106,7 +146,7 @@ public class MinScoreScorerTests extends LuceneTestCase {
 
                         @Override
                         public float matchCost() {
-                            return 10;
+                            return matchCost;
                         }
                     };
                 } else {
@@ -132,23 +172,33 @@ public class MinScoreScorerTests extends LuceneTestCase {
                 return Float.MAX_VALUE;
             }
         };
+        final ScoreMode scoreMode = RandomPicks.randomFrom(
+            random(),
+            new ScoreMode[] { ScoreMode.COMPLETE, ScoreMode.TOP_SCORES, ScoreMode.TOP_DOCS_WITH_SCORES }
+        );
+        final Scorer assertingScorer = AssertingScorer.wrap(random(), scorer, scoreMode);
+        if (twoPhase && randomBoolean()) {
+            return hideTwoPhaseIterator(assertingScorer);
+        } else {
+            return assertingScorer;
+        }
+    }
+
+    private static int[] randomDocs(int maxDoc, int numDocs) {
+        final List<Integer> docs = randomSubsetOf(numDocs, IntStream.range(0, maxDoc).boxed().collect(Collectors.toList()));
+        return docs.stream().mapToInt(n -> n).sorted().toArray();
     }
 
     public void doTestRandom(boolean twoPhase) throws IOException {
         final int maxDoc = TestUtil.nextInt(random(), 10, 10000);
-        final int numDocs = TestUtil.nextInt(random(), 1, maxDoc / 2);
+        final int numDocs = TestUtil.nextInt(random(), 1, maxDoc);
         final Set<Integer> uniqueDocs = new HashSet<>();
         while (uniqueDocs.size() < numDocs) {
             uniqueDocs.add(random().nextInt(maxDoc));
         }
-        final int[] docs = new int[numDocs];
-        int i = 0;
-        for (int doc : uniqueDocs) {
-            docs[i++] = doc;
-        }
-        Arrays.sort(docs);
+        final int[] docs = randomDocs(maxDoc, numDocs);
         final float[] scores = new float[numDocs];
-        for (i = 0; i < numDocs; ++i) {
+        for (int i = 0; i < numDocs; ++i) {
             scores[i] = random().nextFloat();
         }
         Scorer scorer = scorer(maxDoc, docs, scores, twoPhase);
@@ -192,5 +242,49 @@ public class MinScoreScorerTests extends LuceneTestCase {
         for (int iter = 0; iter < iters; ++iter) {
             doTestRandom(true);
         }
+    }
+
+    public void testConjunction() throws Exception {
+        final int maxDoc = randomIntBetween(10, 10000);
+        final Map<Integer, Integer> matchedDocs = new HashMap<>();
+        final List<Scorer> scorers = new ArrayList<>();
+        final int numScorers = randomIntBetween(2, 10);
+        for (int s = 0; s < numScorers; s++) {
+            final int numDocs = randomIntBetween(2, maxDoc);
+            final int[] docs = randomDocs(maxDoc, numDocs);
+            final float[] scores = new float[numDocs];
+            for (int i = 0; i < numDocs; ++i) {
+                scores[i] = randomFloat();
+            }
+            final boolean useTwoPhase = randomBoolean();
+            final Scorer scorer = scorer(maxDoc, docs, scores, useTwoPhase);
+            final float minScore;
+            if (randomBoolean()) {
+                minScore = randomFloat();
+                MinScoreScorer minScoreScorer = new MinScoreScorer(scorer.getWeight(), scorer, minScore);
+                scorers.add(minScoreScorer);
+            } else {
+                scorers.add(scorer);
+                minScore = 0.0f;
+            }
+            for (int i = 0; i < numDocs; i++) {
+                if (scores[i] >= minScore) {
+                    matchedDocs.compute(docs[i], (k, v) -> v == null ? 1 : v + 1);
+                }
+            }
+        }
+
+        final DocIdSetIterator disi = ConjunctionDISI.intersectScorers(scorers);
+        final List<Integer> actualDocs = new ArrayList<>();
+        while (disi.nextDoc() != DocIdSetIterator.NO_MORE_DOCS) {
+            actualDocs.add(disi.docID());
+        }
+        final List<Integer> expectedDocs = matchedDocs.entrySet()
+            .stream()
+            .filter(v -> v.getValue() == numScorers)
+            .map(Map.Entry::getKey)
+            .sorted()
+            .collect(Collectors.toList());
+        assertThat(actualDocs, equalTo(expectedDocs));
     }
 }

